src/plugins/discover/public/components/discover_grid/get_render_cell_value.tsx

```diff
@@ -187,3 +118,89 @@ export const getRenderCellValueFn =
       />
     );
   };
+
+/**
+ * Helper function to show top level objects
+ * this is used for legacy stuff like displaying products of our ecommerce dataset
+ */
+function getInnerColumns(fields: Record<string, unknown[]>, columnId: string) {
+  return Object.fromEntries(
+    Object.entries(fields).filter(([key]) => {
+      return key.indexOf(`${columnId}.`) === 0;
+    })
+  );
+}
+
+/**
+ * Helper function for the cell popover
+ */
+function renderPopoverContent(
+  rowRaw: ElasticSearchHit,
+  rowFlattened: Record<string, unknown>,
+  field: DataViewField | undefined,
+  columnId: string,
+  dataView: DataView,
+  useTopLevelObjectColumns: boolean
+) {
+  if (useTopLevelObjectColumns || field?.type === '_source') {
+    const json = useTopLevelObjectColumns
+      ? getInnerColumns(rowRaw.fields as Record<string, unknown[]>, columnId)
+      : rowRaw;
+    return (
+      <JsonCodeEditor json={json as Record<string, unknown>} width={defaultMonacoEditorWidth} />
+    );
+  }
+
+  return (
+    <span
+      // formatFieldValue guarantees sanitized values
+      // eslint-disable-next-line react/no-danger
+      dangerouslySetInnerHTML={{
+        __html: formatFieldValue(rowFlattened[columnId], rowRaw, dataView, field),
+      }}
+    />
+  );
+}
+/**
+ * Helper function to show top level objects
+ * this is used for legacy stuff like displaying products of our ecommerce dataset
+ */
+function getTopLevelObjectPairs(
+  row: ElasticSearchHit,
+  columnId: string,
+  dataView: DataView,
+  fieldsToShow: string[]
+) {
+  const innerColumns = getInnerColumns(row.fields as Record<string, unknown[]>, columnId);
+  // Put the most important fields first
+  const highlights: Record<string, unknown> = (row.highlight as Record<string, unknown>) ?? {};
+  const highlightPairs: Array<[string, string]> = [];
+  const sourcePairs: Array<[string, string]> = [];
+  Object.entries(innerColumns).forEach(([key, values]) => {
+    const subField = dataView.getFieldByName(key);
+    const displayKey = dataView.fields.getByName
+      ? dataView.fields.getByName(key)?.displayName
+      : undefined;
+    const formatter = subField
+      ? dataView.getFormatterForField(subField)
+      : { convert: (v: unknown, ...rest: unknown[]) => String(v) };
+    const formatted = (values as unknown[])
+      .map((val: unknown) =>
+        formatter.convert(val, 'html', {
+          field: subField,
+          hit: row,
+          indexPattern: dataView,
+        })
+      )
+      .join(', ');
+    const pairs = highlights[key] ? highlightPairs : sourcePairs;
+    if (displayKey) {
```

Comment:
I know I wrote this code, but omg is it horrible 😅

Reply:
so I've got good news for your, since I refactured the structure of the file, but did not touch the function's logic ... now technically: I wrote this code :)